CenterForOpenScience/ember-osf-web
# [ENG-4377] Google dataset discovery

-   Ticket: https://openscience.atlassian.net/browse/ENG-4377
-   Git branch: feature/google-dataset-discovery

## Purpose

The purpose of these changes is to return a Structured Data JSON-LD object to provide Google with a head element tag to be crawled and indexed by its bots. The aim of these changes is to improve the overall crawl rate and with the indexing and ranking of OSF resources.

## Summary of Changes

-Added the returnStructuredData() method and the ScriptTagAttrs type to the Registrations Overview route to return the JSON-LD object from the `osf.io/<guid>/metadata/?format=google-dataset-json-ld` endpoint.
-Added ScriptTagAttrs to the exported interface and updated the Content type to include 'object' for the meta-tags service.
-Added error handling and messaging strings to the translations file when failing to return the object from the endpoint.
-Recommend: renaming the meta-tags service to the HeadTagsService to reference the script and link tags also included by the service.

## Screenshot(s)

<img width="1680" alt="Registration - localhost" src="https://user-images.githubusercontent.com/82047646/232605670-919266d3-e3ef-49f0-879e-b4d22d4e0963.png">

## Side Effects

On different staging environments, the code will need to be re-tested to ensure the proper endpoint is being reached for the parent guid. It will fail if the flag is not set for the endpoint.

## QA Notes

-Use DevTools to verify that the head element script tag displays the proper MIME type attribute as "application/ld+json" and has a valid JSON object as its content. e.g. <head><script type='application/ld+json' src='osf.io/...'>{jsonObject}<script/></head>.
-Is the script tag in its current implementation secure and efficient? Are there any attributes you would suggest adding to the tag to ensure further web application security? `<async>` comes to mind and a more complete list for consideration may be found here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script.
-Mobile checks may not be necessary, but adding the view-source: prefix to the URL as in view-source:https://osf.io/<guid>/ would be a good check.
-Does the JSON object pass both a formatting test and this Google test? Links for each: https://jsonlint.com/ and https://search.google.com/test/rich-results. If you have any questions, please ask!
